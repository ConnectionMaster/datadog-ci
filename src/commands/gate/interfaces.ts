import {Writable} from 'stream'

import {AxiosPromise, AxiosResponse} from 'axios'

import {SpanTags} from '../../helpers/interfaces'

export interface Payload {
  spanTags: SpanTags
}

export interface APIHelper {
  evaluateGateRules(evaluateRequest: Payload, write: Writable['write']): AxiosPromise<AxiosResponse>
}
