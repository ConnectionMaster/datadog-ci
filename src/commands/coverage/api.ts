import fs from 'fs'
import path from 'path'
import {createGzip} from 'zlib'

import type {AxiosPromise, AxiosRequestConfig, AxiosResponse} from 'axios'

import FormData from 'form-data'
import {v4 as uuidv4} from 'uuid'

import {getRequestBuilder} from '../../helpers/utils'

import {Flush, Payload} from './interfaces'

// Dependency follows-redirects sets a default maxBodyLength of 10 MB https://github.com/follow-redirects/follow-redirects/blob/b774a77e582b97174813b3eaeb86931becba69db/index.js#L391
// We don't want any hard limit enforced by the CLI, the backend will enforce a max size by returning 413 errors.
const maxBodyLength = Infinity

export const datadogSite = process.env.DATADOG_SITE || process.env.DD_SITE || 'datadoghq.com'
export const intakeUrl = `https://ci-intake.${datadogSite}`
export const apiUrl = `https://api.${datadogSite}`

export const uploadCodeCoverageReport = (request: (args: AxiosRequestConfig) => AxiosPromise<AxiosResponse>) => async (
  codeCoverageReport: Payload
) => {
  const form = new FormData()

  let fileName
  try {
    fileName = path.parse(codeCoverageReport.path).name
  } catch (e) {
    fileName = 'default_file_name'
  }

  const custom: Record<string, any> = {
    ...codeCoverageReport.spanTags,
    ...codeCoverageReport.customTags,
    ...codeCoverageReport.customMeasures,
    '_dd.hostname': codeCoverageReport.hostname,
    '_dd.report_name': fileName,
    type: 'coverage_report',
    format: codeCoverageReport.format,
  }

  form.append('event', JSON.stringify(custom), {filename: 'event.json'})
  form.append('code_coverage_report_file', fs.createReadStream(codeCoverageReport.path).pipe(createGzip()), {
    filename: `${uuidv4()}.gz`,
  })

  return request({
    data: form,
    headers: form.getHeaders(),
    maxBodyLength,
    method: 'POST',
    url: 'api/v2/cicovreprt',
  })
}

export const flushCodeCoverage = (request: (args: AxiosRequestConfig) => AxiosPromise<AxiosResponse>) => async (
  flushSignal: Flush
) => {
  const form = new FormData()

  const custom: Record<string, any> = {
    ...flushSignal.spanTags,
    ...flushSignal.customTags,
    ...flushSignal.customMeasures,
    '_dd.hostname': flushSignal.hostname,
    type: 'coverage_report',
    flush: 'true',
  }

  form.append('event', JSON.stringify(custom), {filename: 'event.json'})

  return request({
    data: form,
    headers: form.getHeaders(),
    maxBodyLength,
    method: 'POST',
    url: 'api/v2/cicovreprt',
  })
}

export const apiConstructor = (baseIntakeUrl: string, apiKey: string) => {
  const requestIntake = getRequestBuilder({baseUrl: baseIntakeUrl, apiKey})

  return {
    uploadCodeCoverageReport: uploadCodeCoverageReport(requestIntake),
    flushCodeCoverage: flushCodeCoverage(requestIntake),
  }
}
