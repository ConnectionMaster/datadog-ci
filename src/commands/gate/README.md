# GATE command

Evaluate Quality Gates rules in Datadog.

## Usage

### Commands

#### `evaluate`

The `evaluate` command evaluates the matching Quality Gates rules in Datadog.

```bash
datadog-ci gate evaluate [--scope] [--tags] [--dry-run] [--fail-if-unavailable] [--fail-on-empty]
```

For example:

```bash
datadog-ci gate evaluate --scope team:backend --scope team:frontend --fail-on-empty
```

- `--scope` is an array of key value pairs of the form `key:value`. This parameter sets additional scope when retrieving matching rules. Only the rules matching the scope provided will be evaluated.
- `--tags` is an array of key value pairs of the form `key:value`. This parameter sets global tags applied to all results. The upload process merges the tags passed on the command line with the tags in the `DD_TAGS` environment variable. If a key appears in both `--tags` and `DD_TAGS`, the value in `DD_TAGS` takes precedence.
- `--dry-run` (default: `false`): runs the command without the final evaluation step. All other checks are performed.
- `--fail-if-unavailable` (default: `false`): fails the command if there is unavailability on the Datadog side.
- `--fail-on-empty` (default: `false`): fails the command if no matching rules were found in Datadog.

### Environment variables

Additionally, you may configure the `gate` command with environment variables:

- `DATADOG_API_KEY` or `DD_API_KEY` (**required**): API key used to authenticate the requests.
- `DATADOG_APP_KEY` (**required**): APP key used to authenticate the requests.
- `DD_TAGS`: Set global tags applied to all spans. The format must be `key1:value1,key2:value2`. The upload process merges the tags passed on the command line with the tags in the `--tags` parameter. If a key appears in both `--tags` and `DD_TAGS`, the value in `DD_TAGS` takes precedence.
- `DATADOG_SITE`: choose your Datadog site, for example, datadoghq.com or datadoghq.eu.

### Optional dependencies

- [`git`](https://git-scm.com/downloads) is used for extracting repository metadata.

### End-to-end testing process

To verify the command works as expected, use `--dry-run`:

```bash
export DATADOG_API_KEY='<API key>'
export DATADOG_APP_KEY='<APP key>'

yarn launch gate evaluate --scope team:backend --dry-run
```

Successful output looks like the example below:

```bash
ℹ️ Evaluating rules matching the following information:
Repository: git@github.com:DataDog/datadog-ci.git
Branch: master
team: backend

Dry run mode is enabled. Not evaluating the rules.
```
