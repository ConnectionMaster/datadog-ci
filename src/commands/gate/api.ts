import {Writable} from 'stream'

import {AxiosPromise, AxiosRequestConfig, AxiosResponse} from 'axios'

import {getRequestBuilder} from '../../helpers/utils'

import {Payload} from './interfaces'

export const evaluateGateRules = (request: (args: AxiosRequestConfig) => AxiosPromise<AxiosResponse>) => async (
  evaluateRequest: Payload,
  write: Writable['write']
) => {
  // Converting the payload to the format expected from Rapid
  const payload = JSON.stringify({
    data: {
      type: 'gate_evaluation',
      attributes: {
        tags: evaluateRequest.spanTags,
      },
    },
  })

  return request({
    data: payload,
    headers: {
      'Content-Type': 'application/json',
    },
    method: 'POST',
    url: '/api/v2/quality-gates/evaluate',
  })
}

export const apiConstructor = (baseIntakeUrl: string, apiKey: string, appKey: string) => {
  const serviceRequest = getRequestBuilder({baseUrl: baseIntakeUrl, apiKey, appKey})

  return {
    evaluateGateRules: evaluateGateRules(serviceRequest),
  }
}
