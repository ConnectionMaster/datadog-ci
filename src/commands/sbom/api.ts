import {AxiosPromise, AxiosRequestConfig, AxiosResponse} from 'axios'

import {CONTENT_TYPE_HEADER, CONTENT_TYPE_VALUE_PROTOBUF, METHOD_POST} from '../../constants'
import {getRequestBuilder} from '../../helpers/utils'

import {getBaseUrl} from '../junit/utils'

import {API_ENDPOINT, INTAKE_NAME} from './constants'
import {ScaRequest} from './types'

const maxBodyLength = Infinity

/**
 * Get the function to upload our results to the intake.
 * @param apiKey
 */
export const getApiHelper = (apiKey: string): ((scaRequest: ScaRequest) => AxiosPromise<AxiosResponse>) => {
  /**
   * function used to marshall and send the data
   * @param request - the AXIOS element used to send the request
   */
  const uploadSBomPayload = (request: (args: AxiosRequestConfig) => AxiosPromise<AxiosResponse>) => async (
    scaPayload: ScaRequest
  ) => {
    // const buffer = SBOMPayload.encode(payload).finish()

    return request({
      data: JSON.stringify(scaPayload),
      headers: {
        [CONTENT_TYPE_HEADER]: CONTENT_TYPE_VALUE_PROTOBUF,
        'DD-EVP-ORIGIN': 'datadog-ci',
        'DD-EVP-ORIGIN-VERSION': '0.0.1',
      },
      maxBodyLength,
      method: METHOD_POST,
      url: API_ENDPOINT,
    })
  }

  // Get the intake name
  const url = getBaseUrl() + 'api/v2/static-analysis-sca/dependencies'
  // Get the AXIOS request/response function
  const requestIntake = getRequestBuilder({baseUrl: url, apiKey})

  return uploadSBomPayload(requestIntake)
}
