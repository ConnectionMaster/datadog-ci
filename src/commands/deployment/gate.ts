import {isAxiosError} from 'axios'
import chalk from 'chalk'
import {Command, Option} from 'clipanion'
import * as t from 'typanion'

import {FIPS_ENV_VAR, FIPS_IGNORE_ERROR_ENV_VAR} from '../../constants'
import {toBoolean} from '../../helpers/env'
import {enableFips} from '../../helpers/fips'
import {ICONS} from '../../helpers/formatting'
import {Logger, LogLevel} from '../../helpers/logger'
import {getApiHostForSite} from '../../helpers/utils'

import {apiConstructor} from './api'
import {APIHelper, GateEvaluationRequest} from './interfaces'

/**
 * This command allows to evaluate a deployment gate in Datadog.
 * It handles the entire process of requesting a gate evaluation and polling for results
 * The command will exit with status 0 when the gate passes and status 1 otherwise.
 */
export class DeploymentGateCommand extends Command {
  public static paths = [['deployment', 'gate']]

  public static usage = Command.Usage({
    category: 'CI Visibility',
    description: 'Evaluate deployment gates in Datadog.',
    details: `
      This command allows to evaluate a deployment gate in Datadog.
      The command will exit with status 0 when the gate passes and status 1 otherwise.
    `,
    examples: [
      [
        'Evaluate a deployment gate for payments-backend service in prod environment',
        'datadog-ci deployment gate --service payments-backend --env prod',
      ],
      [
        'Evaluate a deployment gate with custom timeout',
        'datadog-ci deployment gate --service payments-backend --env prod --timeout 7200',
      ],
      [
        'Evaluate a deployment gate and fail if an error occurs',
        'datadog-ci deployment gate --service payments-backend --env prod --fail-on-error',
      ],
      [
        'Evaluate a deployment gate with version and APM primary tag',
        'datadog-ci deployment gate --service payments-backend --env prod --version 1.2.3 --apm-primary-tag team:backend',
      ],
    ],
  })

  // Required parameters
  private service = Option.String('--service', {
    description: 'The service name (e.g. payments-backend)',
    validator: t.isString(),
  })
  private env = Option.String('--env', {
    description: 'The environment name (e.g. prod, staging)',
    validator: t.isString(),
  })

  // Optional parameters
  private identifier = Option.String('--identifier', 'default', {
    description: 'The deployment identifier (defaults to "default")',
  })
  private version = Option.String('--version', {
    description: 'The deployment version (required for gates with faulty deployment detection rules)',
  })
  private apmPrimaryTag = Option.String('--apm-primary-tag', {
    description: 'The APM primary tag (only for gates with faulty deployment detection rules)',
  })
  private timeout = Option.String('--timeout', '10800', {
    description: 'Maximum time to wait for the script execution in seconds (default: 10800 = 3 hours)',
    validator: t.isString(),
  })
  private failOnError = Option.Boolean('--fail-on-error', false, {
    description:
      'When true, the script will consider the gate as failed when timeout is reached or unexpected errors occur calling the Datadog APIs',
  })

  // FIPS options
  private fips = Option.Boolean('--fips', false)
  private fipsIgnoreError = Option.Boolean('--fips-ignore-error', false)

  private config = {
    apiKey: process.env.DATADOG_API_KEY || process.env.DD_API_KEY,
    appKey: process.env.DATADOG_APP_KEY || process.env.DD_APP_KEY,
    fips: toBoolean(process.env[FIPS_ENV_VAR]) ?? false,
    fipsIgnoreError: toBoolean(process.env[FIPS_IGNORE_ERROR_ENV_VAR]) ?? false,
  }

  private logger: Logger = new Logger((s: string) => this.context.stdout.write(s), LogLevel.INFO)
  private pollingInterval = 15000
  private startTime: number = Date.now()

  public async execute() {
    enableFips(this.fips || this.config.fips, this.fipsIgnoreError || this.config.fipsIgnoreError)

    if (!this.service) {
      this.logger.error(chalk.red(`${ICONS.FAILED} Missing required parameter: ${chalk.bold('--service')}`))

      return 1
    }

    if (!this.env) {
      this.logger.error(chalk.red(`${ICONS.FAILED} Missing required parameter: ${chalk.bold('--env')}`))

      return 1
    }

    const timeoutSeconds = parseInt(this.timeout, 10)
    if (isNaN(timeoutSeconds) || timeoutSeconds <= 0) {
      this.logger.error(
        chalk.red(`${ICONS.FAILED} Invalid ${chalk.bold('--timeout')} value. Must be a positive integer.`)
      )

      return 1
    }

    if (!this.config.apiKey) {
      this.logger.error(
        `${ICONS.FAILED} Neither ${chalk.red.bold('DATADOG_API_KEY')} nor ${chalk.red.bold(
          'DD_API_KEY'
        )} are in your environment.`
      )

      return 1
    }

    if (!this.config.appKey) {
      this.logger.error(
        `${ICONS.FAILED} Neither ${chalk.red.bold('DATADOG_APP_KEY')} nor ${chalk.red.bold(
          'DD_APP_KEY'
        )} are in your environment.`
      )

      return 1
    }

    this.logger.info('Starting deployment gate evaluation with parameters:')
    this.logger.info(`\tService: ${this.service}`)
    this.logger.info(`\tEnvironment: ${this.env}`)
    this.logger.info(`\tIdentifier: ${this.identifier}`)
    if (this.version) {
      this.logger.info(`\tVersion: ${this.version}`)
    }
    if (this.apmPrimaryTag) {
      this.logger.info(`\tAPM Primary Tag: ${this.apmPrimaryTag}`)
    }
    this.logger.info(`\tTimeout: ${timeoutSeconds} seconds`)
    this.logger.info(`\tFail on error: ${this.failOnError ? 'true' : 'false'}\n`)

    try {
      const api = this.getApiHelper(this.config.apiKey, this.config.appKey)
      const evaluationRequest = this.buildEvaluationRequest()

      const evaluationId = await this.requestGateEvaluation(api, evaluationRequest)

      return await this.pollForEvaluationResults(api, evaluationId, timeoutSeconds)
    } catch (error) {
      this.logger.error(`Deployment gate evaluation failed: ${error instanceof Error ? error.message : String(error)}`)

      if (isAxiosError(error)) {
        if (error.response?.status && error.response.status >= 400 && error.response.status < 500) {
          this.logger.error(`${ICONS.FAILED} Request failed with client error, exiting with status 1`)

          return 1
        }
      }

      return this.getExitCodeForDatadogError()
    }
  }

  private getApiHelper(apiKey: string, appKey: string): APIHelper {
    const site = process.env.DATADOG_SITE || process.env.DD_SITE || 'datadoghq.com'
    const baseAPIURL = `https://${getApiHostForSite(site)}`

    return apiConstructor(baseAPIURL, apiKey, appKey)
  }

  private buildEvaluationRequest(): GateEvaluationRequest {
    const request: GateEvaluationRequest = {
      service: this.service || '',
      env: this.env || '',
      identifier: this.identifier,
    }

    if (this.version) {
      request.version = this.version
    }

    if (this.apmPrimaryTag) {
      request.apm_primary_tag = this.apmPrimaryTag
    }

    return request
  }

  private async requestGateEvaluation(api: APIHelper, request: GateEvaluationRequest): Promise<string> {
    this.logger.info('Requesting gate evaluation...')

    try {
      const response = await api.requestGateEvaluation(request)
      const evaluationId = response.data.data.attributes.evaluation_id
      this.logger.info(chalk.green(`Gate evaluation started successfully. Evaluation ID: ${evaluationId}\n`))

      return evaluationId
    } catch (error) {
      if (isAxiosError(error) && error.response?.status) {
        this.logger.error(`Request failed with client error: ${error.response.status} ${error.response.statusText}`)
      } else {
        this.logger.error(
          `Could not start gate evaluation with unknown error: ${
            error instanceof Error ? error.message : String(error)
          }`
        )
      }

      return Promise.reject(error)
    }
  }

  private async pollForEvaluationResults(
    api: APIHelper,
    evaluationId: string,
    timeoutSeconds: number
  ): Promise<number> {
    this.logger.info('Waiting for gate evaluation results...')

    const maxWaitTime = timeoutSeconds * 1000

    while (true) {
      const timePassed = Date.now() - this.startTime
      const remainingTime = maxWaitTime - timePassed

      if (remainingTime <= 0) {
        this.logger.warn(
          `${ICONS.WARNING} Timeout reached (${timeoutSeconds} seconds). Gate evaluation did not complete.`
        )

        return this.getExitCodeForDatadogError()
      }

      try {
        const response = await api.getGateEvaluationResult(evaluationId)
        const status = response.data.data.attributes.gate_status

        switch (status) {
          case 'pass':
            this.logger.info(chalk.green(`\t${ICONS.SUCCESS} Gate evaluation passed`))

            return 0
          case 'fail':
            this.logger.info(chalk.red(`\t${ICONS.FAILED} Gate evaluation failed`))

            return 1
          case 'in_progress':
            const waitTime = Math.min(this.pollingInterval, remainingTime)
            const waitTimeInSeconds = Math.floor(waitTime / 1000)
            const rules = response.data.data.attributes.rules
            const totalRules = rules.length
            const completedRules = rules.filter((rule) => rule.status !== 'in_progress').length

            this.logger.info(
              `\tGate evaluation in progress (${completedRules}/${totalRules} rules completed). Retrying in ${waitTimeInSeconds}s...`
            )

            await new Promise((resolve) => setTimeout(resolve, waitTime))
            continue
          default:
            throw new Error(`Unknown gate evaluation status: ${status}`)
        }
      } catch (error) {
        this.logger.error(
          `Error polling for gate evaluation results: ${error instanceof Error ? error.message : String(error)}`
        )

        return this.getExitCodeForDatadogError()
      }
    }
  }

  private getExitCodeForDatadogError() {
    if (this.failOnError) {
      this.logger.info('Unexpected error happened, exiting with status 1')

      return 1
    }

    this.logger.info('Unexpected error happened, exiting with status 0')

    return 0
  }
}
