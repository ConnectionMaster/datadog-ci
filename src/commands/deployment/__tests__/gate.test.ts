import {createCommand, makeRunCLI} from '../../../helpers/__tests__/testing-tools'

import * as apiModule from '../api'
import {DeploymentGateCommand} from '../gate'

describe('gate', () => {
  describe('execute', () => {
    const runCLI = makeRunCLI(DeploymentGateCommand, ['deployment', 'gate'], {skipResetEnv: true})

    let originalEnv: NodeJS.ProcessEnv

    beforeEach(() => {
      originalEnv = {...process.env}
      process.env.DATADOG_API_KEY = 'test-api-key'
      process.env.DATADOG_APP_KEY = 'test-app-key'

      jest.useFakeTimers()
    })

    afterEach(() => {
      process.env = originalEnv
      jest.useRealTimers()
      jest.clearAllMocks()
    })

    test('should fail if service is not provided', async () => {
      const {context, code} = await runCLI(['--env', 'prod'])
      expect(code).toBe(1)
      expect(context.stdout.toString()).toContain('Missing required parameter: --service')
    })

    test('should fail if env is not provided', async () => {
      const {context, code} = await runCLI(['--service', 'test-service'])
      expect(code).toBe(1)
      expect(context.stdout.toString()).toContain('Missing required parameter: --env')
    })

    test('should fail if API key is not provided', async () => {
      delete process.env.DATADOG_API_KEY
      delete process.env.DD_API_KEY

      const {context, code} = await runCLI(['--service', 'test-service', '--env', 'prod'])
      expect(code).toBe(1)
      expect(context.stdout.toString()).toContain('Neither DATADOG_API_KEY nor DD_API_KEY are in your environment')
    })

    test('should fail if APP key is not provided', async () => {
      delete process.env.DATADOG_APP_KEY
      delete process.env.DD_APP_KEY

      const {context, code} = await runCLI(['--service', 'test-service', '--env', 'prod'])
      expect(code).toBe(1)
      expect(context.stdout.toString()).toContain('Neither DATADOG_APP_KEY nor DD_APP_KEY are in your environment')
    })

    test('should fail if timeout is invalid', async () => {
      const {context, code} = await runCLI(['--service', 'test-service', '--env', 'prod', '--timeout', 'invalid'])
      expect(code).toBe(1)
      expect(context.stdout.toString()).toContain('Invalid --timeout value. Must be a positive integer.')
    })

    test('should succeed when gate evaluation passes on first poll', async () => {
      const mockApi = {
        requestGateEvaluation: jest.fn().mockResolvedValue({
          data: {
            data: {
              attributes: {
                evaluation_id: 'test-evaluation-id',
              },
            },
          },
        }),
        getGateEvaluationResult: jest.fn().mockResolvedValue({
          data: {
            data: {
              attributes: {
                gate_status: 'pass',
              },
            },
          },
        }),
      }

      const apiConstructorSpy = jest.spyOn(apiModule, 'apiConstructor').mockReturnValue(mockApi)

      const {context, code} = await runCLI(['--service', 'test-service', '--env', 'prod'])

      expect(code).toBe(0)
      expect(context.stdout.toString()).toContain('Starting deployment gate evaluation')
      expect(context.stdout.toString()).toContain('Requesting gate evaluation...')
      expect(context.stdout.toString()).toContain(
        'Gate evaluation started successfully. Evaluation ID: test-evaluation-id'
      )
      expect(context.stdout.toString()).toContain('Waiting for gate evaluation results...')
      expect(context.stdout.toString()).toContain('Gate evaluation passed')

      expect(apiConstructorSpy).toHaveBeenCalledWith('https://app.datad0g.com', 'test-api-key', 'test-app-key')
      expect(mockApi.requestGateEvaluation).toHaveBeenCalledTimes(1)
      expect(mockApi.requestGateEvaluation).toHaveBeenCalledWith({
        service: 'test-service',
        env: 'prod',
        identifier: 'default',
      })
      expect(mockApi.getGateEvaluationResult).toHaveBeenCalledTimes(1)
      expect(mockApi.getGateEvaluationResult).toHaveBeenCalledWith('test-evaluation-id')
    })

    test('should fail when gate evaluation fails after in_progress', async () => {
      const mockApi = {
        requestGateEvaluation: jest.fn().mockResolvedValue({
          data: {
            data: {
              attributes: {
                evaluation_id: 'test-evaluation-id',
              },
            },
          },
        }),
        getGateEvaluationResult: jest
          .fn()
          .mockResolvedValueOnce({
            data: {
              data: {
                attributes: {
                  gate_status: 'in_progress',
                },
              },
            },
          })
          .mockResolvedValueOnce({
            data: {
              data: {
                attributes: {
                  gate_status: 'fail',
                },
              },
            },
          }),
      }

      const apiConstructorSpy = jest.spyOn(apiModule, 'apiConstructor').mockReturnValue(mockApi)

      const runPromise = runCLI(['--service', 'test-service', '--env', 'prod'])

      // Run all timers and wait for all pending promises to resolve
      await jest.runAllTimersAsync()

      const {context, code} = await runPromise

      expect(code).toBe(1)
      expect(context.stdout.toString()).toContain('Starting deployment gate evaluation')
      expect(context.stdout.toString()).toContain('Requesting gate evaluation...')
      expect(context.stdout.toString()).toContain(
        'Gate evaluation started successfully. Evaluation ID: test-evaluation-id'
      )
      expect(context.stdout.toString()).toContain('Waiting for gate evaluation results...')
      expect(context.stdout.toString()).toContain('Gate evaluation failed')

      expect(apiConstructorSpy).toHaveBeenCalledWith('https://app.datad0g.com', 'test-api-key', 'test-app-key')
      expect(mockApi.requestGateEvaluation).toHaveBeenCalledTimes(1)
      expect(mockApi.requestGateEvaluation).toHaveBeenCalledWith({
        service: 'test-service',
        env: 'prod',
        identifier: 'default',
      })
      expect(mockApi.getGateEvaluationResult).toHaveBeenCalledTimes(2)
      expect(mockApi.getGateEvaluationResult).toHaveBeenNthCalledWith(1, 'test-evaluation-id')
      expect(mockApi.getGateEvaluationResult).toHaveBeenNthCalledWith(2, 'test-evaluation-id')
    })
  })

  describe('buildEvaluationRequest', () => {
    test('should build basic request with required parameters', () => {
      const command = createCommand(DeploymentGateCommand)
      command['service'] = 'test-service'
      command['env'] = 'prod'
      command['identifier'] = 'default'

      const request = command['buildEvaluationRequest']()
      expect(request).toEqual({
        service: 'test-service',
        env: 'prod',
        identifier: 'default',
      })
    })

    test('should include version when provided', () => {
      const command = createCommand(DeploymentGateCommand)
      command['service'] = 'test-service'
      command['env'] = 'prod'
      command['identifier'] = 'default'
      command['version'] = '1.2.3'

      const request = command['buildEvaluationRequest']()
      expect(request).toEqual({
        service: 'test-service',
        env: 'prod',
        identifier: 'default',
        version: '1.2.3',
      })
    })

    test('should include apm_primary_tag when provided', () => {
      const command = createCommand(DeploymentGateCommand)
      command['service'] = 'test-service'
      command['env'] = 'prod'
      command['identifier'] = 'default'
      command['apmPrimaryTag'] = 'team:backend'

      const request = command['buildEvaluationRequest']()
      expect(request).toEqual({
        service: 'test-service',
        env: 'prod',
        identifier: 'default',
        apm_primary_tag: 'team:backend',
      })
    })

    test('should include both version and apm_primary_tag when provided', () => {
      const command = createCommand(DeploymentGateCommand)
      command['service'] = 'test-service'
      command['env'] = 'prod'
      command['identifier'] = 'default'
      command['version'] = '1.2.3'
      command['apmPrimaryTag'] = 'team:backend'

      const request = command['buildEvaluationRequest']()
      expect(request).toEqual({
        service: 'test-service',
        env: 'prod',
        identifier: 'default',
        version: '1.2.3',
        apm_primary_tag: 'team:backend',
      })
    })
  })
})
