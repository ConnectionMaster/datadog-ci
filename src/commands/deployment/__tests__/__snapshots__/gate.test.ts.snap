// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`gate execute evaluation errors on gate evaluation request should fail when gate evaluation request fails with 500 and fail-on-error is true 1`] = `
"Starting deployment gate evaluation with parameters:
	Service: test-service
	Environment: prod
	Identifier: default
	Timeout: 10800 seconds
	Fail on error: true

Requesting gate evaluation...
Request failed with client error: 500 Internal Server Error
Deployment gate evaluation failed: Request failed with status code 500
Unexpected error happened, exiting with status 1
"
`;

exports[`gate execute evaluation errors on gate evaluation request should pass when gate evaluation request fails with 500 1`] = `
"Starting deployment gate evaluation with parameters:
	Service: test-service
	Environment: prod
	Identifier: default
	Timeout: 10800 seconds
	Fail on error: false

Requesting gate evaluation...
Request failed with client error: 500 Internal Server Error
Deployment gate evaluation failed: Request failed with status code 500
Unexpected error happened, exiting with status 0
"
`;

exports[`gate execute evaluation errors on gate evaluation result pass with a 500 error 1`] = `
"Starting deployment gate evaluation with parameters:
	Service: test-service
	Environment: prod
	Identifier: default
	Timeout: 10800 seconds
	Fail on error: false

Requesting gate evaluation...
Gate evaluation started successfully. Evaluation ID: test-evaluation-id

Waiting for gate evaluation results...
Error polling for gate evaluation results: API Error
Unexpected error happened, exiting with status 0
"
`;

exports[`gate execute evaluation errors on gate evaluation result should fail with 500 error when fail-on-error is true 1`] = `
"Starting deployment gate evaluation with parameters:
	Service: test-service
	Environment: prod
	Identifier: default
	Timeout: 10800 seconds
	Fail on error: true

Requesting gate evaluation...
Gate evaluation started successfully. Evaluation ID: test-evaluation-id

Waiting for gate evaluation results...
Error polling for gate evaluation results: Request failed with status code 500
Unexpected error happened, exiting with status 1
"
`;

exports[`gate execute evaluation errors on gate evaluation result should handle invalid evaluation status 1`] = `
"Starting deployment gate evaluation with parameters:
	Service: test-service
	Environment: prod
	Identifier: default
	Timeout: 10800 seconds
	Fail on error: false

Requesting gate evaluation...
Gate evaluation started successfully. Evaluation ID: test-evaluation-id

Waiting for gate evaluation results...
Error polling for gate evaluation results: Unknown gate evaluation status: expired
Unexpected error happened, exiting with status 0
"
`;

exports[`gate execute successful evaluation should fail when gate evaluation fails 1`] = `
"Starting deployment gate evaluation with parameters:
	Service: test-service
	Environment: prod
	Identifier: default
	Timeout: 10800 seconds
	Fail on error: false

Requesting gate evaluation...
Gate evaluation started successfully. Evaluation ID: test-evaluation-id

Waiting for gate evaluation results...
	❌ Gate evaluation failed
"
`;

exports[`gate execute successful evaluation should pass when gate evaluation passes after multiple in_progress calls 1`] = `
"Starting deployment gate evaluation with parameters:
	Service: test-service
	Environment: prod
	Identifier: default
	Timeout: 10800 seconds
	Fail on error: false

Requesting gate evaluation...
Gate evaluation started successfully. Evaluation ID: test-evaluation-id

Waiting for gate evaluation results...
	Gate evaluation in progress (1/1 rules completed). Retrying in 15s...
	Gate evaluation in progress (1/1 rules completed). Retrying in 15s...
	✅ Gate evaluation passed
"
`;

exports[`gate execute successful evaluation should succeed when gate evaluation passes on first poll 1`] = `
"Starting deployment gate evaluation with parameters:
	Service: test-service
	Environment: prod
	Identifier: default
	Timeout: 10800 seconds
	Fail on error: false

Requesting gate evaluation...
Gate evaluation started successfully. Evaluation ID: test-evaluation-id

Waiting for gate evaluation results...
	✅ Gate evaluation passed
"
`;
